createTheme throwing type errors with typescript

createTheme is throwing compile time error 

<!--
  Thank you very much for contributing to Material-UI by creating an issue! ❤️
  To avoid duplicate issues we ask you to check off the following list.
-->

<!-- Checked checkbox should look like this: [x] -->

- [x] The issue is present in the latest release.
- [ ] I have searched the [issues](https://github.com/mui-org/material-ui/issues) of this repository and believe that this is not a duplicate.

## Current Behavior 😯

``` javascript
node_modules/@material-ui/core/styles/createTheme.d.ts:10:18 - error TS2430: Interface 'import("/Users/vrajole/sg-core/frontend-monorepo/packages/lib/auth/node_modules/@material-ui/core/styles/createTheme").ThemeOptions' incorrectly extends interface 'import("/Users/vrajole/sg-core/frontend-monorepo/packages/lib/auth/node_modules/@material-ui/system/createTheme/createTheme").ThemeOptions'.
  Types of property 'zIndex' are incompatible.
    Type 'Partial<ZIndex> | undefined' is not assignable to type 'Record<string, number> | undefined'.
      Type 'Partial<ZIndex>' is not assignable to type 'Record<string, number>'.
        Property 'mobileStepper' is incompatible with index signature.
          Type 'number | undefined' is not assignable to type 'number'.
            Type 'undefined' is not assignable to type 'number'.

10 export interface ThemeOptions extends SystemThemeOptions {
                    ~~~~~~~~~~~~


Found 1 error.

error Command failed with exit code 1.
```

## Expected Behavior 🤔

<!-- Describe what should happen. -->

## Steps to Reproduce 🕹

<!--
  Provide a link to a live example (you can use codesandbox.io) and an unambiguous set of steps to reproduce this bug.
  Include code to reproduce, if relevant (which it most likely is).

  You should use the official codesandbox template as a starting point: https://material-ui.com/r/issue-template-next

  If you have an issue concerning TypeScript please start from this TypeScript playground: https://material-ui.com/r/ts-issue-template

  Issues without some form of live example have a longer response time.
-->

Steps:

1. Use material ui mentioned version with typescript (No need to create new theme)

## Context 🔦

<!--
  What are you trying to accomplish? How has this issue affected you?
  Providing context helps us come up with a solution that is most useful in the real world.
-->

## Your Environment 🌎

```
@material-ui/core: 5.0.0-alpha.37,
@material-ui/icons: 5.0.0-alpha.37,
Node: 15.8.0 
Yarn: 1.22.4
npm: 7.17.0 
```

<!--
  Run `npx @material-ui/envinfo` and post the results.
  If you encounter issues with TypeScript please include the used tsconfig.
-->
<details>
  <summary>`npx @material-ui/envinfo`</summary>
  
```
  Don't forget to mention which browser you used.
  Output from `npx @material-ui/envinfo` goes here.
```
</details>

@mnajdova To reproduce, clone https://github.com/fmeum/mui-reproducer and run
```
yarn install && npx tsc
```
to get the output:
```
node_modules/@mui/material/styles/createTheme.d.ts:10:18 - error TS2430: Interface 'import("/home/fhenneke/git/mui-reproducer/node_modules/@mui/material/styles/createTheme").ThemeOptions' incorrectly extends interface 'import("/home/fhenneke/git/mui-reproducer/node_modules/@mui/system/createTheme/createTheme").ThemeOptions'.
  Types of property 'zIndex' are incompatible.
    Type 'Partial<ZIndex> | undefined' is not assignable to type 'Record<string, number> | undefined'.
      Type 'Partial<ZIndex>' is not assignable to type 'Record<string, number>'.
        Property 'mobileStepper' is incompatible with index signature.
          Type 'number | undefined' is not assignable to type 'number'.
            Type 'undefined' is not assignable to type 'number'.

10 export interface ThemeOptions extends SystemThemeOptions {
                    ~~~~~~~~~~~~


Found 1 error.
```

I can't reproduce this anymore if I do one of the following:
- Update typescript to major version 4.
- Disable `strictNullChecks` in the `tsconfig.json`.

--- a/packages/mui-material/src/styles/createTheme.d.ts
+++ b/packages/mui-material/src/styles/createTheme.d.ts
@@ -7,7 +7,7 @@ import { Transitions, TransitionsOptions } from './createTransitions';
 import { ZIndex, ZIndexOptions } from './zIndex';
 import { Components } from './components';
 
-export interface ThemeOptions extends SystemThemeOptions {
+export interface ThemeOptions extends Omit<SystemThemeOptions, 'zIndex'> {
   mixins?: MixinsOptions;
   components?: Components;
   palette?: PaletteOptions;

